feat/sk-gen code review

hmerrilees opened the pull request:
- [x] I certify that this PR does not contain any code that has been generated with GitHub Copilot or any other AI-based code generation tool, in accordance with this project's policies.

Commit:
working... check out sk-gen

Commit:
cleanup interface

Commit:
main description

Commit:
don't include pod lifecycle

Commit:
spelling

Commit:
exactly trace length, as opposed to at minimum trace length

Commit:
fixed length arg

Commit:
fixed display flag

Commit:
next step comments

hmerrilees committed:
Merge branch 'acrlabs:master' into feat/synthetic-graph

Commit:
feat: :sparkles: create/delete deployments, action-based search, graphviz output

Graph states are now discovered by enumerating actions at a depth of trace_length - 1.

Commit:
feat: :sparkles: separated graph output into its own subcommand, allowing for much larger graphs to be generated in the same amount of time owing to not needing to waste any time enumerating (the much more computationally complex) traces.

Commit:
pre-simplification

Commit:
post-simplification

Commit:
feat: inlined trace length validation

baltaazr committed:
feat: walks with sampling

baltaazr committed:
fix

baltaazr committed:
fix

baltaazr committed:
remove

baltaazr committed:
fix

Commit:
grouped cli args by function

Commit:
minor doc improvements and link fixes

Commit:
comment improvements

Commit:
simplify create_deployment

Commit:
fix: Added output.rs

hmerrilees commented on sk-gen/make_sim.sh:
Not sure how this one made it in... but here it is I guess.

baltaazr commented:
LGTM! 🚀

WalkerKarina committed:
support multinode start

sayaks1 commented on sk-gen/src/main.rs:
```diff
@@ -0,0 +1,700 @@
+//!
+//! [`TraceEvent`] represents the Kubernetes API call which corresponds to a `ClusterAction`.
+//!
+//! [`Edge`] stores both a `ClusterAction` and the corresponding `TraceEvent`.
```
Why is ClusterAction not a property of TraceEvent?

hmerrilees replied:
We don't have control over TraceEvent, which is the atomic unit of change ingested by simkube. ClusterAction is internal to our generation, we cache the TraceEvent on the edge so that we don't have to regenerate every walk. Other approaches may be better suited to that task.

WalkerKarina commented on sk-gen/src/main.rs:
```diff
@@ -0,0 +1,700 @@
+        let walk_start_node = self.graph.node_indices().next().unwrap();
+
+        // We use a depth-first search because eventually we may want to use stochastic methods which do not
+        // fully enumerate the neighborhood of each visited node.
```
Is this implying that eventually we want to no longer produce complete graphs but only generate certain neighbors of a surrounding region?

hmerrilees replied:
Yes! Potentially. This goes along with your other comment of not generating all the actions.

WalkerKarina commented on sk-gen/src/main.rs:
```diff
@@ -0,0 +1,700 @@
+        }
+    }
+
+    /// Enumerates at least all possible `ClusterAction` instances.
```
This makes sense but wondering if it is efficient enough - do we anticipate performance issues of computing all possible and then eliminating vs eliminating on the fly?

hmerrilees replied:
Potentially! And this isolation of generation/execution makes it very easy to make requisite changes to that end!

Commit:
add initial trace

hmerrilees committed:
Merge pull request #2 from ACRL-Clinic-24-25/sk-gen-kk-saya

support multinode start

Commit:
fixed unwrap none on hard-coded trace

Commit:
containers

Commit:
wip resource request

Commit:
containers + memory resource requests

Commit:
use simkube tracestore (unfortunately we lose json output for now)

baltaazr committed:
feat: add GPU resource type

MaxistheSpy committed:
fix: include target folder in gitignore

previously target folder was not included in gitingore resulting in
compiled rust files being tracked by git

MaxistheSpy committed:
doc: minor rewording of ClusterAction docs

MaxistheSpy committed:
feat: added cpu resources and finailized initial resource action implementation

This commit finalizes the implementation for cpu resources
gpu resources now follow step policy rather than having seperate
increment decrement functions
gpu resources are now bounds checked
fixed gpu resources not being included in json ouput
gpu resources assumed to be nvidia
removed gpu_model option for now
overall consolidated the codebase to use simmilar conventions

Note: theres a somewhat significant naming issue here, to "request" a
gpu your actually supposed to put it in "limits" this is a weird api
call on the kube side, we could put it in request part of the json but
its bad form and im not 100% sure it works. so our gpu resources being
stored in the requests struct of the container mismatches the json
structure. that being said it is a resource request so its not entirely
off base to put it there.

MaxistheSpy committed:
added sensible configurable defaults for resources
